Simplify forward / backward AD for linalg.eigh and add checks

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70528
* #70527

This PR adds checks for the backward of `linalg.eigh`, similar to those
deduced in https://github.com/pytorch/pytorch/pull/70253

It also makes its the implementation parallel that of the (fwd/bwd) derivative of
`torch.linalg.eig` and it makes most OpInfo tests pass.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

Differential Revision: [D33530149](https://our.internmc.facebook.com/intern/diff/D33530149)

lezcano committed:
Correct forward / backward AD for linalg.eigh

This PR adds checks for the backward of `linalg.eigh`, similar to those
deduced in https://github.com/pytorch/pytorch/pull/70253

It also makes its the implementation parallel that of the (fwd/bwd) derivative of
`torch.linalg.eig` and it makes most OpInfo tests pass.

[ghstack-poisoned]

lezcano committed:
Update on "Simplify forward / backward AD for linalg.eigh and add checks"


This PR adds checks for the backward of `linalg.eigh`, similar to those
deduced in https://github.com/pytorch/pytorch/pull/70253

It also makes its the implementation parallel that of the (fwd/bwd) derivative of
`torch.linalg.eig` and it makes most OpInfo tests pass.

[ghstack-poisoned]

IvanYashchuk reviewed (commented):
For real symmetric matrices, the resulting eigenvectors are real and they are normalized but their sign is not unique, so the loss invariance problem is not only for the complex case.

Lezcano commented:
In the real case it is fine since, even though you can multiply the matrix by `-1` you still have a "locally unique" decomposition.

Using the words of differential geometry, you have a principal bundle with discrete fibres (think of something like a covering space). In this case, the total space and the base space are locally isometric, so you may do the computations on the total space and life is good.

Lezcano commented:
In other words, there's not much we can do in the real case to check for that invariance as, since the fibres are discrete (think of the set {1, -1} vs {e^{i\phi} | \phi \in R}}, we cannot use the gradient information to look into these things, because gradients and discrete things don't play well together.

IvanYashchuk commented on torch/linalg/__init__.py:
```diff
@@ -602,13 +602,11 @@
-             the subspace by a rotation matrix and the resulting eigenvectors will be valid
-             eigenvectors.
+             For this reason, the loss function shall not depend on the phase of the eigenvectors, as
+             this quantity is not well-defined. This will be checked when computing the gradients of the loss function.
```
Alright, I think it should note that the check is performed only in the complex case.

lezcano committed:
Update on "Simplify forward / backward AD for linalg.eigh and add checks"


This PR adds checks for the backward of `linalg.eigh`, similar to those
deduced in https://github.com/pytorch/pytorch/pull/70253

It also makes its the implementation parallel that of the (fwd/bwd) derivative of
`torch.linalg.eig` and it makes most OpInfo tests pass.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Simplify forward / backward AD for linalg.eigh and add checks"


This PR adds checks for the backward of `linalg.eigh`, similar to those
deduced in https://github.com/pytorch/pytorch/pull/70253

It also makes its the implementation parallel that of the (fwd/bwd) derivative of
`torch.linalg.eig` and it makes most OpInfo tests pass.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Simplify forward / backward AD for linalg.eigh and add checks"


This PR adds checks for the backward of `linalg.eigh`, similar to those
deduced in https://github.com/pytorch/pytorch/pull/70253

It also makes its the implementation parallel that of the (fwd/bwd) derivative of
`torch.linalg.eig` and it makes most OpInfo tests pass.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano committed:
Update on "Simplify forward / backward AD for linalg.eigh and add checks"


This PR adds checks for the backward of `linalg.eigh`, similar to those
deduced in https://github.com/pytorch/pytorch/pull/70253

It also makes its the implementation parallel that of the (fwd/bwd) derivative of
`torch.linalg.eig` and it makes most OpInfo tests pass.

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

@IvanYashchuk do you want to do take a final look as well?

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33530149).

IvanYashchuk reviewed (approved):
Looks good!
